hilti/runtime/include/util.h

```diff
@@ -649,4 +650,97 @@ constexpr std::size_t hashCombine(std::size_t hash1, std::size_t hash2, Hashes..
         return result;
 }
 
+template<typename Data, typename Error>
+class ControlBlockRef;
+
+// Helper class for loosely tracking liveliness of some memory. Code using this
+// would instantiate an instance of this pointing to some data. It can then
+// pass out references to that control block by calling `asRef`. One can get
+// values from these references as long as the original control block is live.
+template<typename Data, typename Error>
+class ControlBlock {
+public:
+    using Ref = ControlBlockRef<Data, Error>;
+
+    ControlBlock() = default;
+    ControlBlock(Data* data) : _data(data) {}
+
+    ControlBlock(const ControlBlock&) = default;
+    ControlBlock(ControlBlock&&) = default;
+
+    ControlBlock& operator=(const ControlBlock&) = default;
+    ControlBlock& operator=(ControlBlock&&) = default;
+
+    friend bool operator==(const ControlBlock& a, const ControlBlock& b) {
+        return std::tie(a._control, a._data) == std::tie(b._control, b._data);
+    }
+
+    friend bool operator!=(const ControlBlock& a, const ControlBlock& b) { return ! (a == b); }
+
+    /*implicit*/ operator Ref() const {
+        if ( ! _control )
+            _control = std::make_shared<bool>();
+
+        return {_control, _data};
+    }
+
+    // Invalidate all references to this control block.
+    void reset() { _control.reset(); }
```

Comment:
Should this clear `_data` as well, just to avoid accidental continued use? (Or is it explicitly supported to create a new control pointer for the same data?)

Reply:
Setting `_data` here does nothing since all access to it is via a reference instance which store their own value. I believe for an expired reference it should be impossible to get access to the value behind that `_data` since all access to data is guarded by checks for the reference's `_control`.